# 01 — Inicijalni pregled skupa podataka

U ovoj svesci prvi put zavirujemo u sirove podatke. Ideja je da se upoznamo sa strukturom skupa, pogledamo tipove podataka i uočimo eventualne probleme (nedostajuće vrednosti, duplikate, besmislene vrednosti) pre nego što bilo šta diramo.

Bitna napomena: ovde podatke samo posmatramo i dijagnostikujemo, ne menjamo ih. Čišćenje i obrada dolaze u kasnijim sveskama, tek nakon podele na trening i test skup.

## 1. Učitavanje podataka i biblioteka

Koristimo pandas, standardni alat za rad sa tabelarnim podacima u Python-u. Skup učitavamo iz CSV fajla u DataFrame — u suštini tabelu sa redovima i kolonama.

In [ ]:
#Prvi put importujemo biblioteku numpy (np) - za numeričke operacije; koristićemo ga kasnije za rad sa NaN vrednostima.
import pandas as pd
import numpy as np

# Podešavanje da pandas prikazuje SVE kolone kada ispisujemo tabelu.
# Bez ovoga, pandas skraćuje prikaz i sakriva srednje kolone sa "...".
pd.set_option("display.max_columns", None)


df = pd.read_csv("../WA_Fn-UseC_-Telco-Customer-Churn.csv")

print("Podaci su uspešno učitani.")

Podaci su uspešno učitani.


## 2. Dimenzije skupa podataka

Prvo proveravamo osnovne dimenzije — koliko redova (korisnika) i koliko kolona
(atributa) skup sadrži.

In [ ]:
print("Broj redova (korisnika):", df.shape[0])
print("Broj kolona (atributa):", df.shape[1])

Broj redova (korisnika): 7043
Broj kolona (atributa): 21


## 3. Prvi pogled na podatke

Prikazujemo prvih nekoliko redova kako bismo videli konkretne vrednosti i stekli
osećaj za sadržaj svake kolone.

In [ ]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 4. Informacije o kolonama i tipovima podataka

`info()` nam daje pregled svih kolona: naziv, broj popunjenih (ne-null) vrednosti i tip podatka (`dtype`).

Ovo je zgodno za rano hvatanje problema. Najviše nas zanima da li se tipovi poklapaju sa onim što očekujemo — recimo, `TotalCharges` bi trebalo da bude broj, pa ako je učitana kao tekst (`object`), to je već znak da nešto nije kako treba.

In [4]:
# df.info() ispisuje pregled: naziv kolone, broj ne-null vrednosti i tip podataka.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## 5. Provera nedostajućih vrednosti



In [ ]:
# df.isnull() vraća tabelu True/False (True = vrednost nedostaje).
# .sum() sabira po kolonama: koliko True (nedostajućih) ima u svakoj koloni.
print("Broj nedostajućih vrednosti po kolonama:")
print(df.isnull().sum())

Broj nedostajućih vrednosti po kolonama:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


## 6. Skrivene nedostajuće vrednosti

Prethodna provera je pokazala da nema nedostajućih vrednosti, ali to zna da prevari. U praksi se nedostajuće vrednosti nekad kriju kao prazni razmaci (`" "`) umesto pravih `NaN` oznaka. `isnull()` ih ne uhvati jer su tehnički gledano tekst, a ne prazne vrednosti.

Zato dodatno proveravamo da li neka kolona sadrži vrednosti koje su samo prazan razmak.

In [6]:
# Poredimo svaku ćeliju sa praznim razmakom " ".
# (df == " ") vraća tabelu True/False (True = ćelija sadrži tačno jedan razmak).
# .sum() sabira po kolonama koliko takvih "praznih" vrednosti ima.
print("Broj vrednosti koje su prazan razmak, po kolonama:")
print((df == " ").sum())

Broj vrednosti koje su prazan razmak, po kolonama:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


## 7. Analiza problema sa kolonom `TotalCharges`

Sad je jasno zašto je `TotalCharges` prepoznata kao tekst (`object`) umesto broja: pošto 11 vrednosti sadrži prazan razmak (a to nije broj), pandas je celu kolonu protumačio kao tekstualnu.


In [ ]:
# pd.to_numeric() pokušava da pretvori kolonu u broj.
# errors="coerce" znači: vrednosti koje se NE mogu pretvoriti u broj postaju NaN umesto da izazovu grešku.
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

# Ponovo proveravamo nedostajuće vrednosti — sada bi TotalCharges trebalo da pokaže 11 NaN vrednosti.
print("Broj nedostajućih vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())

Broj nedostajućih vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


Zaključak je da `TotalCharges` treba da bude numerička kolona, pa je konvertujemo.

In [ ]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"],errors="coerce")

## 8. Analiza korisnika sa nedostajućim `TotalCharges`

Sad kad je `TotalCharges` numerička, možemo da vidimo tačno koliko korisnika ima nedostajuću vrednost. Hoćemo i da razumemo zašto baš oni — da li je posredi slučajna greška ili tu postoji neki obrazac.

In [9]:
print("Nedostajuće vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())
print()


Nedostajuće vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64



### Ključna opservacija

Svih 11 korisnika sa nedostajućom vrednošću `TotalCharges` ima `tenure = 0`. To su, dakle, korisnici koji su tek počeli da koriste uslugu, pa u trenutku formiranja skupa za njih još nije bilo evidentirane ukupne naplate.

Ovo nam dosta govori. Nije u pitanju slučajna greška u merenju niti namerno preskočeno polje — vrednost jednostavno logički ne postoji za te korisnike (nova pretplata, naplate još nije bilo).

Šta ćemo sa ovim redovima, odlučićemo kasnije, nakon podele na trening i test skup, da ne bismo napravili curenje informacija (data leakage).

Zaključak je da nedostajuće vrednosti ipak postoje u skupu, pa u sledećem koraku utvrđujemo koje su tačno.

In [ ]:
print("Korisnici sa nedostajućim TotalCharges:")
print(df[df["TotalCharges"].isnull()])
print()

Korisnici sa nedostajućim TotalCharges:
      customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service        

## 9. Provera duplikata

Proveravamo da li skup sadrži duplirane redove, kao i da li se neki `customerID`
pojavljuje više puta. Duplikati mogu iskriviti kasnije analize i modeliranje.

In [ ]:
print("Broj dupliranih redova:", df.duplicated().sum())
print("Broj dupliranih customerID vrednosti:",df["customerID"].duplicated().sum())
print()


Broj dupliranih redova: 0
Broj dupliranih customerID vrednosti: 0



Zakljucujemo da `nema duplikata` u nasem skupu podataka

## 10. Osnovna statistika numeričkih promenljivih

Za numeričke kolone (`SeniorCitizen`, `tenure`, `MonthlyCharges`, `TotalCharges`) `describe()` daje osnovne pokazatelje: broj vrednosti, prosek, standardnu devijaciju, minimum, maksimum i kvartile.

Ovo nam koristi da vidimo u kom se rasponu vrednosti kreću (npr. najveća/najmanja mesečna naplata), da uhvatimo eventualne besmislice (negativne vrednosti tamo gde ih ne sme biti) i da steknemo grubu sliku o raspodeli pre detaljnije analize u narednim sveskama.

In [12]:
print("Osnovna statistika numeričkih promenljivih:")
print(df[["SeniorCitizen", "tenure","MonthlyCharges", "TotalCharges"]].describe())
print()

Osnovna statistika numeričkih promenljivih:
       SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7043.000000  7043.000000     7043.000000   7032.000000
mean        0.162147    32.371149       64.761692   2283.300441
std         0.368612    24.559481       30.090047   2266.771362
min         0.000000     0.000000       18.250000     18.800000
25%         0.000000     9.000000       35.500000    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.850000   3794.737500
max         1.000000    72.000000      118.750000   8684.800000



### Tumačenje osnovne statistike

Par stvari koje se odmah vide:

- `tenure` ide od 0 do 72 meseca (znači do 6 godina). Ona nula se odnosi na onih 11 korisnika sa nedostajućim `TotalCharges` koje smo malopre našli.
- `MonthlyCharges` se kreće od oko 18 do oko 119 dolara, bez negativnih i besmislenih vrednosti.
- `TotalCharges` ide od oko 18 do oko 8.685 dolara. Broj vrednosti je 7032 umesto 7043 jer 11 redova ima `NaN`, što se poklapa sa prethodnom analizom.
- `SeniorCitizen` je prikazana kao broj (0 i 1), ali je zapravo binarna kategorija, pa `describe()` nad njom nema mnogo smisla — bolja je provera raspodele kategorija.

## 11. Provera kategorijskih promenljivih

### 11.1 SeniorCitizen — binarna kategorija u brojnom obliku

`SeniorCitizen` je zapisana kao broj (0 ili 1), ali je suštinski binarna kategorija: da li je korisnik stariji građanin ili ne. Zato umesto statistike gledamo raspodelu po vrednostima, da vidimo koliko korisnika pada u svaku grupu.

In [13]:
print("Raspodela SeniorCitizen:")
print(df["SeniorCitizen"].value_counts())
print()

Raspodela SeniorCitizen:
SeniorCitizen
0    5901
1    1142
Name: count, dtype: int64



Većina korisnika (~5.900, oko 84%) nisu stariji građani, dok ih je manjina (~1.140, oko 16%) u toj grupi. Ova nesrazmera će nam biti bitna kasnije — proverićemo da li starost korisnika ima veze sa odlivom.

Izdvajamo sve kolone koje imaju tekstualne vrednosti.

Za svaku kategoriju prikazujemo sve *`postojece`* kategorije i *`broj`* korisnika u svakoj kategoriji

### 11.2 Sve kategorijske promenljive

Prolazimo kroz sve kolone tekstualnog tipa (`object`) i za svaku ispisujemo postojeće kategorije zajedno sa brojem korisnika. Tako vidimo tačno koje se vrednosti javljaju, da li ima nekih neočekivanih ili nekonzistentnih kategorija i koliko su kategorije izbalansirane (da li negde jedna vrednost dominira).

`customerID` preskačemo — tih 7043 jedinstvenih identifikatora nema šta da nam kaže u ovakvom prikazu.

In [16]:
# Izdvajamo sve kolone tekstualnog tipa.
categorical_columns = df.select_dtypes(include="object").columns

# Za svaku kategorijsku kolonu prikazujemo raspodelu.
# Izuzimamo customerID jer predstavlja jedinstveni identifikator korisnika,
# a ne kategoriju za analizu.
for column in categorical_columns:
    if column == "customerID":
        continue
    print(f"\n{column}:")
    print(df[column].value_counts())


gender:
gender
Male      3555
Female    3488
Name: count, dtype: int64

Partner:
Partner
No     3641
Yes    3402
Name: count, dtype: int64

Dependents:
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

PhoneService:
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

MultipleLines:
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64

InternetService:
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

OnlineSecurity:
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64

OnlineBackup:
OnlineBackup
No                     3088
Yes                    2429
No internet service    1526
Name: count, dtype: int64

DeviceProtection:
DeviceProtection
No                     3095
Yes                    2422
No internet service    1526
Name: count, dtype: int64

TechSupport:
TechSupport
No   

### Tumačenje kategorijskih promenljivih

**Balansirane binarne kolone:**
- `gender`: Male 3555, Female 3488 — praktično 50-50
- `Partner`: No 3641, Yes 3402 — malo više onih bez partnera

**Nebalansirane binarne kolone:**
- `Dependents`: No 4933, Yes 2110 — većina nema zavisne osobe
- `PhoneService`: Yes 6361, No 682 — skoro svi imaju telefonsku uslugu
- `PaperlessBilling`: Yes 4171, No 2872

**Kolone sa tri vrednosti (`Yes` / `No` / `"No internet service"` i sl.**
`MultipleLines`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`.

`"No internet service"` i `"No"` nisu isto:
`"No"` — korisnik ima internet, ali ne koristi tu dodatnu uslugu
`"No internet service"` — korisnik uopšte nema internet
Ova razlika će biti bitna za feature engineering, pa ćemo o tome (spajati ih ili ne) odlučiti kasnije.

**Kolone sa više kategorija:**
- `InternetService`: Fiber optic 3096, DSL 2416, No 1520
- `Contract`: Month-to-month 3875 (ubedljivo najviše), One year 1472, Two year 1685
- `PaymentMethod`: Electronic check 2365, Mailed check 1604, Bank transfer 1542, Credit card 1521

**Ciljna promenljiva `Churn`:**
- No: 5173 (73%)
- Yes: 1869 (27%)

Klase su nebalansirane, što će direktno uticati na izbor metrika i na strategiju modeliranja.

---

## 12. Zaključak

Kroz inicijalni pregled skupa `Telco Customer Churn` došli smo do sledećeg.

**Struktura skupa:** 7.043 korisnika i 21 kolona, koje smo podelili u 4 grupe — demografske, usluge, podaci o nalogu i plaćanju, i ciljna promenljiva.

**Problemi koje ćemo rešavati:**
1. Skrivene nedostajuće vrednosti u `TotalCharges` — 11 korisnika (0.16%) sa `tenure = 0`. Posle konverzije u broj ti redovi imaju `NaN`. Šta ćemo sa njima, odlučujemo tek nakon podele na trening/test.
2. `SeniorCitizen` je zapisana kao 0/1 iako je binarna kategorija. Razmislićemo o konverziji u `No`/`Yes` radi konzistentnosti.
3. Vrednosti `"No internet service"` i `"No phone service"` nisu isto što i `"No"`. O njihovom tretmanu odlučujemo u fazi pripreme podataka.
4. Nebalansirane klase u `Churn` (~73% No, ~27% Yes), što dosta utiče na izbor metrika i na pristup modeliranju.

**Ono što je u redu:**
- Nema duplikata (ni potpunih redova ni po `customerID`)
- Numeričke kolone nemaju nelogične vrednosti (negativne ili van opsega)
- Kategorijske kolone imaju konzistentne vrednosti, bez tipfelera